### PR TITLE
Implement ETag

### PR DESCRIPTION
**Pull request by Danielius1922**

### CODE REVIEW

File: api/unittest/etagtest.cpp [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+{
+  EXPECT_EQ(CODE, data->code);
+  *static_cast<bool *>(data->user_data) = true;
+  oc::TestDevice::Terminate();
+  OC_DBG("GET payload: %s", oc::RepPool::GetJson(data->payload).data());
+}
+
+TEST_F(TestETagWithServer, GetResourceWithETag)
+{
+  oc_log_set_level(OC_LOG_LEVEL_DEBUG);
```
Comment (Danielius1922) [inline api/unittest/etagtest.cpp:354]: rm

File: api/oc_etag.c [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  uint64_t now = oc_clock_time();
+  if (now > g_etag) {
+    g_etag = now;
+  }
+  g_etag += etag_random();
+  if (g_etag == 0) {
+    // 0 is reserved for no etag
+    g_etag = 1;
+  }
+  return g_etag;
```
Comment (jkralik) [inline api/oc_etag.c:70]: add TODO to check overflow

File: api/oc_server_api.c [hunk 3]
```diff
@@ -863,4 +874,30 @@ oc_notify_observers_delayed_ms(oc_resource_t *resource, uint16_t milliseconds)
   oc_notify_observers_delayed_ticks(resource, ticks);
 }
 
+void
+oc_notify_resource_changed(oc_resource_t *resource)
+{
+  assert(resource != NULL);
+  oc_notify_observers(resource);
```
Comment (jkralik) [inline api/oc_server_api.c:881]: before invoke `oc_notify_observers` we need to check whether resource has flag observable

File: CMakeLists.txt [hunk 4]
```diff
@@ -60,6 +60,8 @@ set(OC_INOUT_BUFFER_SIZE "" CACHE STRING "Custom static buffer size for network
 set(OC_INOUT_BUFFER_POOL "" CACHE STRING "Custom static pool size of network messages.")
 set(OC_APP_DATA_BUFFER_SIZE "" CACHE STRING "Custom static buffer size for application messages.")
 set(OC_APP_DATA_BUFFER_POOL "" CACHE STRING "Custom static size of application messages.")
+# set(OC_ETAG_ENABLED OFF CACHE BOOL "Enable Entity Tag (ETag) support.")
+set(OC_ETAG_ENABLED ON CACHE BOOL "Enable Entity Tag (ETag) support.")
```
Comment (jkralik) [inline CMakeLists.txt:64]: Default OFF

File: api/oc_etag.c [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  }
+}
+
+static int
+etag_store_encode(size_t device, void *data)
+{
+  (void)data;
+  oc_rep_start_root_object();
+  etag_iterate_resources(device, etag_iterate_encode_resource, NULL);
+  oc_rep_end_root_object();
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply api/oc_etag.c:151]: I think we can keep addressing by URI, that's unlikely to change and to keep the possibility of the representation being changed we can use an object.

```json
{
  "<uri>": {
    "etag": <etag>,
     ...
  } 
}

```

File: api/oc_ri.c [hunk 6]
```diff
@@ -1440,14 +1282,52 @@ oc_ri_invoke_coap_entity_handler(const coap_packet_t *request,
...
+     * ETag in the header. If so, and if the ETag matches the ETag of the
+     * resource, then the response will be a 2.03 response with an empty
+     * payload.
+     */
+    uint64_t etag = 0;
+    const uint8_t *etag_buf = NULL;
+    size_t etag_buf_len = coap_options_get_etag(request, &etag_buf);
+    if (etag_buf_len == sizeof(etag)) {
+      memcpy(&etag, etag_buf, etag_buf_len);
+      not_modified = (etag != 0 && etag == cur_resource->etag);
```
Comment (jkralik) [inline api/oc_ri.c:1316]: `tag != 0`changed to `tag != OC_ETAG_UNINITALIZED` or something similar.

File: api/oc_etag.c [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+{
+  uint64_t now = oc_clock_time();
+  if (now > g_etag) {
+    g_etag = now;
+  }
+  g_etag += etag_random();
+  if (g_etag == OC_ETAG_UNINITALIZED) {
+    g_etag = 1;
+  }
+  return g_etag;
```
Comment (jkralik) [inline api/oc_etag.c:70]: just add here: `TODO: check for overflow` - all resources need to reinitialize etags.

File: api/oc_etag.c [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    *rep_data->etag = (uint64_t)etag;
+  }
+  return 0;
+}
+
+static int
+etag_iterate_clear_etag(oc_resource_t *resource, void *data)
+{
+  (void)data;
+  oc_resource_set_etag(resource, 0);
```
Comment (jkralik) [inline api/oc_etag.c:235]: `OC_ETAG_UNINITALIZED`

File: api/oc_etag.c [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    if (ret <= 0) {
+      OC_ERR("failed to load etag for device %zu", i);
+      success = false;
+    }
+  }
+  g_etag = etag + etag_random();
+  if (g_etag == OC_ETAG_UNINITALIZED) {
+    g_etag = 1;
+  }
+  OC_DBG("g_tag: %" PRIu64, g_etag);
```
Comment (jkralik) [inline api/oc_etag.c:283]: also add `TODO: check flow overflow`

**Pull request merged**